### PR TITLE
format diff msg output for xcat inventory

### PR DESCRIPTION
**Pull request by xuweibj**
#84 

code based on #83

UT:
* json
```
# xcat-inventory diff -o hardware_admin/xcat-inventory/c910f03c17k21/cluster.json -n /tmp/cluster.json
====================BEGIN=====================


--- hardware_admin/xcat-inventory/c910f03c17k21/cluster.json
+++ /tmp/cluster.json
{
    "node": {
        "c910f03c17k10": {
            "network_info": {
                "primarynic": {
                    "mac": [
-                        "42:6e:0a:03:11:0a",
-                        "42:73:0a:03:11:0a!*NOIP*",
+                        "42:6e:0a:03:11:04"
                    ]
                }
            }
        },
-        "f6u13": {
-            "obj_info": {
-                "groups": "openbmc"
-            },
-            "role": "compute",
-            "device_type": "server",
-            "engines": {
-                "hardware_mgt_engine": {
-                    "engine_info": {
-                        "bmc": "10.6.13.100",
-                        "bmcpassword": "0penBmc",
-                        "bmcusername": "root"
-                    },
-                    "engine_type": "openbmc"
-                }
-            },
-            "obj_type": "node"
-        }
+        "boston_ipmi11111": {
+            "obj_info": {
+                "groups": "ipmi"
+            },
+            "role": "compute",
+            "device_type": "server",
+            "engines": {
+                "hardware_mgt_engine": {
+                    "engine_info": {
+                        "bmc": "9.3.23.42"
+                    },
+                    "engine_type": "ipmi"
+                }
+            },
+            "obj_type": "node"
+        }
        "boston_ipmi": {
            "engines": {
                "hardware_mgt_engine": {
                    "engine_info": {
-                        "bmcpassword": "ADMIN",
-                        "bmcusername": "ADMIN"
                    }
                }
            }
        },
        "f6u03": {
            "engines": {
                "hardware_mgt_engine": {
                    "engine_info": {
-                        "bmcpassword": "0penBmc"
+                        "bmcpassword": "password"
                    }
                }
            }
        }
    },
-    "route": {},
-    "schema_version": "1.0"
    "passwd": {
-        "redfish": {
-            "username": "ADMIN",
-            "password": "ADMIN"
-        }
    }
}

====================END=====================
```
* yaml
```
#xcat-inventory diff -o hardware_admin/xcat-inventory/c910f03c17k21/cluster.yaml -n /tmp/cluster.yaml /tmp/cluster.yaml

====================BEGIN=====================


--- hardware_admin/xcat-inventory/c910f03c17k21/cluster.json
+++ /tmp/cluster.json
-route: {}
-schema_version: '1.0'
node:
+  boston_ipmi11111:
+    device_type: server
+    engines:
+      hardware_mgt_engine:
+        engine_info:
+          bmc: 9.3.23.42
+        engine_type: ipmi
+    obj_info:
+      groups: ipmi
+    obj_type: node
+    role: compute
-  f6u13:
-    device_type: server
-    engines:
-      hardware_mgt_engine:
-        engine_info:
-          bmc: 10.6.13.100
-          bmcpassword: 0penBmc
-          bmcusername: root
-        engine_type: openbmc
-    obj_info:
-      groups: openbmc
-    obj_type: node
-    role: compute
  boston_ipmi:
    engines:
      hardware_mgt_engine:
        engine_info:
-          bmcpassword: ADMIN
-          bmcusername: ADMIN
  c910f03c17k10:
    network_info:
      primarynic:
        mac:
-        - 42:6e:0a:03:11:0a
-        - 42:73:0a:03:11:0a!*NOIP*
+        - 42:6e:0a:03:11:04
  f6u03:
    engines:
      hardware_mgt_engine:
        engine_info:
+          bmcpassword: password
-          bmcpassword: 0penBmc
passwd:
-  redfish:
-    password: ADMIN
-    username: ADMIN


====================END=====================

### CODE REVIEW

File: xcat-inventory/xcclient/inventory/invdiff [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+import json
+import sys
+import subprocess
+import re
+
+
+class InvalidFileException(Exception):
+    pass
+
+def runCommand(cmd, env=None):
```
Comment (robin2008) [inline xcat-inventory/xcclient/inventory/invdiff]: Why we need this method here ?

File: xcat-inventory/xcclient/inventory/invdiff [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+import deepdiff
+import pprint
+import yaml
+import json
+import sys
+import subprocess
+import re
+
+
+class InvalidFileException(Exception):
```
Comment (robin2008) [inline xcat-inventory/xcclient/inventory/invdiff]: Put all exception into `exceptions.py`

File: xcat-inventory/xcclient/inventory/invdiff [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+
+class format_diff_output(object):
+
+    def __init__(self, format_type):
+        self.format_type = format_type 
+        self.flag_dict = {'-diff': '-', '+diff': '+'}
+
+    def _get_path_as_list(self, path):
+        path_str = path.replace(']', '')
```
Comment (immarvin) [inline xcat-inventory/xcclient/inventory/invdiff]: Please use `re.findall` to pick out the sub key list, refer to https://www.guru99.com/python-regular-expressions-complete-tutorial.html#7

**Comment by immarvin** [comment]
the diff logic and output is ok. But seems the option and usage is quite confusing.  I think we need more discussion on the user interface before we push this fix in

**Comment by robin2008** [comment]
@immarvin This is a separate function of `xcat-invetory` to  compare two inventory based on file.

Based on this we can extend more capabilities on it,  now it looks good enough.

**Comment by immarvin** [comment]
hi @robin2008 , I think we should be careful when defining user interface. For a two-way comparison,   I do not think we need options like `--origin`, `--new` to specify the object/file s to compare.  `--type git/normal` is quite confusing, I cannot get what this option for from an user's perspective.  

Here is my draft proposal on this: 

the user interface `xcat-inventory diff` will provide the following functions:
1) the comparison between 2 inventory files, can be invoked by `git difftool` to provide a diff view in git
`xcat-inventory diff --from-files <inventory file: A> <inventory file: B>` 

2) the comparison  between the inventory data of 2 xCAT objects in xCATDB
`xcat-inventory diff --from-objs <obj: A> <obj: B>`

3) the comparison between the specified inventory file and inventory data in xCAT DB
`xcat-inventory diff  [--file-vs-obj] <file>`

any comment or suggestion is appreciated

thanks

**Pull request merged**